src/seb/non_hf_models/sonar_model.py

```diff
@@ -0,0 +1,94 @@
+from collections.abc import Sequence
+from pathlib import Path
+from typing import Union
+
+import torch
+from fairseq2.data import Collater
+from fairseq2.data.data_pipeline import read_sequence
+from fairseq2.data.text import TextTokenizer, read_text
+from fairseq2.models.sequence import SequenceBatch
+from fairseq2.typing import Device
+from sonar.inference_pipelines.utils import extract_sequence_batch
+from sonar.models import SonarEncoderModel, SonarEncoderOutput
+from sonar.models.sonar_text import (
+    load_sonar_text_encoder_model,
+    load_sonar_tokenizer,
+)
+
+from seb.model_interface import ModelInterface
+
+
+def truncate_seq_length(
+    sequence_batch: SequenceBatch, max_seq_len: int = 514,
+) -> SequenceBatch:
+    sequence_batch.seqs = sequence_batch.seqs[:, :max_seq_len]
+    sequence_batch.seq_lens = torch.clamp(sequence_batch.seq_lens, max=max_seq_len)  # type: ignore
+    return sequence_batch
+
+
+class SonarTextToEmbeddingModelPipeline(torch.nn.Module, ModelInterface):
+    def __init__(
+        self,
+        encoder: Union[str, SonarEncoderModel],
+        tokenizer: Union[str, TextTokenizer],
+        device: Device = torch.device("cpu"),
+    ) -> None:
+        """
+        Args:
+            encoder (Union[str, SonarEncoderModel]): either cart name or model object
+            tokenizer (Union[str, TextTokenizer]): either cart name or tokenizer object
+            device (device, optional): . Defaults to cpu.
+        """
+        super().__init__()
+        if isinstance(encoder, str):
+            encoder = load_sonar_text_encoder_model(
+                encoder, device=device, progress=False,
+            )
+        if isinstance(tokenizer, str):
+            tokenizer = load_sonar_tokenizer(tokenizer, progress=False)
+
+        self.tokenizer = tokenizer
+        self.model = encoder.to(device).eval()
+        self.device = device
+
+    @torch.inference_mode()
+    def encode(
+        self, input: Union[Path, Sequence[str]], source_lang: str, batch_size: int,
+    ) -> torch.Tensor:
```

Comment:
Hmm it will not have access to the source_lang at inference time

Reply:
An option would be to create 4 instances of this model (one for each language)

Reply:
So the models would be sonar_dan, sonar_nno, etc. (alternatively we can let the encode attribute have access to the task language). However, there are tasks such as language identification where it would get multiple language labels.